### ISSUE TITLE
Zed Preview Crash / Panic, Unsure of cause

### ISSUE THREAD

**Issue by Otterpocket** [zed-industries/community#2392]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Unknown reason causes Zed to crash regularly.

### Environment

Zed: v0.119.4 (Zed Preview)
OS: macOS 14.2.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.
```
.....
2024-01-10T15:35:24 [ERROR] crates/editor/src/display_map/wrap_map.rs:249: no font found
2024-01-10T15:35:24 [ERROR] crates/editor/src/display_map/wrap_map.rs:249: no font found
2024-01-10T15:35:24 [ERROR] crates/editor/src/display_map/wrap_map.rs:249: no font found
2024-01-10T15:35:24 [ERROR] crates/editor/src/display_map/wrap_map.rs:249: no font found
2024-01-10T15:35:24 [ERROR] crates/editor/src/display_map/wrap_map.rs:249: no font found
2024-01-10T15:35:24 [ERROR] crates/editor/src/display_map/wrap_map.rs:249: no font found
2024-01-10T15:35:24 [ERROR] crates/editor/src/display_map/wrap_map.rs:249: no font found
2024-01-10T15:35:24 [ERROR] {
  "thread": "main",
  "payload": "called `Option::unwrap()` on a `None` value",
  "location_data": {
    "file": "crates/editor/src/display_map/block_map.rs",
    "line": 895
  },
  "backtrace": [
    "<backtrace::capture::Backtrace>::create",
    "<backtrace::capture::Backtrace>::new",
    "Zed::init_panic_hook::{closure#0}",
    "std::panicking::rust_panic_with_hook",
    "std::panicking::begin_panic_handler::{{closure}}",
    "std::sys_common::backtrace::__rust_end_short_backtrace",
    "_rust_begin_unwind",
    "core::panicking::panic_fmt",
    "core::panicking::panic",
    "<editor::display_map::block_map::BlockBufferRows as core::iter::traits::iterator::Iterator>::next",
    "<editor::element::EditorElement>::compute_layout::{closure#0}",
    "<editor::element::EditorElement as gpui::element::Element>::paint::{closure#0}",
    "<editor::element::EditorElement as gpui::element::Element>::paint",
    "<gpui::element::DrawableElement<editor::element::EditorElement>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<editor::element::EditorElement>> as gpui::element::ElementObject>::paint",
    "<gpui::element::DrawableElement<gpui::view::AnyView>>::paint::{closure#0}",
    "<gpui::element::DrawableElement<gpui::view::AnyView>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::view::AnyView>> as gpui::element::ElementObject>::paint",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::paint",
    "<gpui::element::DrawableElement<gpui::elements::div::Div>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::elements::div::Div>> as gpui::element::ElementObject>::paint",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::paint",
    "<gpui::element::DrawableElement<gpui::elements::div::Div>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::elements::div::Div>> as gpui::element::ElementObject>::paint",
    "<gpui::element::DrawableElement<gpui::view::View<workspace::status_bar::StatusBar>>>::paint::{closure#0}",
    "<gpui::element::DrawableElement<gpui::view::View<workspace::status_bar::StatusBar>>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::view::View<workspace::status_bar::StatusBar>>> as gpui::element::ElementObject>::paint",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::paint",
    "<gpui::element::DrawableElement<gpui::elements::div::Div>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::elements::div::Div>> as gpui::element::ElementObject>::paint",
    "<gpui::element::DrawableElement<gpui::element::AnyElement>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::element::AnyElement>> as gpui::element::ElementObject>::paint",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::paint",
    "<gpui::element::DrawableElement<gpui::elements::div::Div>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::elements::div::Div>> as gpui::element::ElementObject>::paint",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::paint",
    "<gpui::element::DrawableElement<gpui::elements::div::Div>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::elements::div::Div>> as gpui::element::ElementObject>::paint",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::paint",
    "<gpui::element::DrawableElement<gpui::elements::div::Stateful<gpui::elements::div::Div>>>::paint::{closure#0}",
    "<gpui::window::WindowContext as gpui::window::BorrowWindow>::with_element_state::<gpui::elements::div::DivState, (), <gpui::element::DrawableElement<gpui::elements::div::Stateful<gpui::elements::div::Div>>>::paint::{closure#0}>::{closure#0}",
    "<gpui::element::DrawableElement<gpui::elements::div::Stateful<gpui::elements::div::Div>>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::elements::div::Stateful<gpui::elements::div::Div>>> as gpui::element::ElementObject>::paint",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::paint::<<gpui::elements::div::Div as gpui::element::Element>::paint::{closure#0}>::{closure#0}::{closure#0}::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::paint",
    "<gpui::element::DrawableElement<gpui::elements::div::Div>>::paint",
    "<core::option::Option<gpui::element::DrawableElement<gpui::elements::div::Div>> as gpui::element::ElementObject>::paint",
    "<gpui::view::AnyView>::draw",
    "<gpui::window::WindowContext>::draw",
    "<gpui::window::Window>::new::{closure#0}",
    "gpui::platform::mac::window::display_layer",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run",
    "Zed::main",
    "std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>",
    "std::rt::lang_start::<()>::{closure#0}",
    "std::rt::lang_start_internal",
    "std::rt::lang_start::<()>",
    "_main"
  ],
  "app_version": "0.119.4",
  "release_channel": "Zed Preview",
  "os_name": "macOS",
  "os_version": "14.2.0",
  "architecture": "aarch64",
  "panicked_on": 1704900924765,
  "installation_id": "c8984f04-2978-427a-9d2d-71180d3b1161",
  "session_id": "58ce0f65-96c2-4c0f-8ec6-ab5493acdafe"
}
```

**Comment by ConradIrwin** [comment]
Hi @Otterpocket, thank you for reporting this, and sorry that it's panicking – it should not!

We're struggling to reproduce this issue, and I wondered if you had some time to either reply in-line or screen-share with me at some point today to see if we can figure it out.

* One hypothesis is that the "no font found" error above is a contributing cause. Would you mind sharing your settings, and which font you are using?
* Another hypothesis is that there's something about a specific file you're editing that causes this. Did you notice that to be the case, and if so could you share that file?
* We're seeing some other panics that seem related to this (but our panic handling infrastructure does not let us tell whether you are also the person seeing those). Would you be able to open the `Console` app on macOS (`cmd-space "Console"`) click on "Crash reports", click "reveal in finder" and upload any that start with zed-?.
* * I suspect there are many, if so the best might be to create a zip file and email it to conrad@zed.dev.

Thank you in advance, I look forward to figuring this out!

**Comment by sljmn** [comment]
crashes for me when i select and highlight text in any file.
i cleared the settings.json file restarted zed. opend random file in my project. select some text and it freezes and i need to force restart zedd preview

Environment

Zed Preview 0.119.5
OS: macOS 14.2.1
Memory: 16 GiB
Architecture: aarch64

**Comment by ConradIrwin** [comment]
@sljmn that does not sound good at all!

If Zed is freezing, we likely aren't seeing crash reports. Do you have time to screen share with me to figure it out?

If not, next time it happens, could you please open Activity Monitor and in the "..." menu at the top click "Sample Process" and upload the stack trace. That will help us get to the bottom of what's happening.

Thank you in advance,
Conrad

**Comment by sljmn** [comment]
it happens all the time i can reproduce it often as you want - i will do a screenshare, how to proceed?

**Comment by ConradIrwin** [comment]
If you happen to be around right now, https://meet.google.com/rfw-raqk-fwy?ijlm=1704935026088&adhoc=1&hs=187.

Otherwise please email conrad@zed.dev with a good time for tomorrow!

**Comment by ConradIrwin** [comment]
@Otterpocket Thanks for your email follow-up! I'll take a look. Publishing the salient piece here so everyone can see it:

This seems to be caused by the following config:
```
"language_overrides": {
  "Ruby": {
    "formatter": {
      "external": {
        "command": "standardrb",
        "arguments": ["--fix", "--stdin", "ignored", "--stderr"]
      }
    }
  }
}
```
Editing a `./bin/bundle` (ruby) file causes the crash

**Comment by ConradIrwin** [comment]
@Otterpocket I believe we have fixed your issue in the latest preview (0.119.6, please let me know if you experience otherwise!).

@sljmn we have a fix for your deadlock too; but it didn't make it to 0.119.6, but will be out in the next release (either later today or tomorrow)

**Comment by sljmn** [comment]
using 0.119.8 and it seems fixed! thanks @ConradIrwin

**Comment by ConradIrwin** [comment]
Yes, both should be fixed as of 0.119.8. Thanks for checking in.
